core/capabilities/compute/compute.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	capabilitiespb "github.com/smartcontractkit/chainlink-common/pkg/capabilities/pb"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	coretypes "github.com/smartcontractkit/chainlink-common/pkg/types/core"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/wasm/host"
+	wasmpb "github.com/smartcontractkit/chainlink-common/pkg/workflows/wasm/pb"
+)
+
+const (
+	CapabilityIDCompute = "custom_compute@1.0.0"
+
+	binaryKey = "binary"
+	configKey = "config"
+)
+
+var (
+	computeWASMInit = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "compute_wasm_module_init",
+		Help: "how long it takes to initialize a WASM module",
+		Buckets: []float64{
+			float64(50 * time.Millisecond),
+			float64(100 * time.Millisecond),
+			float64(200 * time.Millisecond),
+			float64(500 * time.Millisecond),
+			float64(1 * time.Second),
+			float64(2 * time.Second),
+			float64(4 * time.Second),
+			float64(8 * time.Second),
+		},
+	}, []string{"workflowID", "stepRef"})
+	computeWASMExec = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "compute_wasm_module_exec",
+		Help: "how long it takes to execute a request from a WASM module",
+		Buckets: []float64{
+			float64(50 * time.Millisecond),
+			float64(100 * time.Millisecond),
+			float64(200 * time.Millisecond),
+			float64(500 * time.Millisecond),
+			float64(1 * time.Second),
+			float64(2 * time.Second),
+			float64(4 * time.Second),
+			float64(8 * time.Second),
+		},
+	}, []string{"workflowID", "stepRef"})
+)
+
+var _ capabilities.ActionCapability = (*Compute)(nil)
+
+type Compute struct {
+	log      logger.Logger
+	registry coretypes.CapabilitiesRegistry
+	modules  *moduleCache
+}
+
+func (c *Compute) RegisterToWorkflow(ctx context.Context, request capabilities.RegisterToWorkflowRequest) error {
+	return nil
+}
+
+func (c *Compute) UnregisterFromWorkflow(ctx context.Context, request capabilities.UnregisterFromWorkflowRequest) error {
+	return nil
+}
+
+func generateID(binary []byte) string {
+	id := sha256.Sum256(binary)
+	return fmt.Sprintf("%x", id)
+}
+
+func (c *Compute) Execute(ctx context.Context, request capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	binary, err := c.popBytesValue(request.Config, binaryKey)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	config, err := c.popBytesValue(request.Config, configKey)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, fmt.Errorf("invalid request: %w", err)
+	}
+
+	id := generateID(binary)
+
+	m, ok := c.modules.get(id)
+	if !ok {
+		mod, err := c.initModule(id, binary, request.Metadata.WorkflowID, request.Metadata.ReferenceID)
```

Comment:
Random thinking: how much time is spent initializing the module?
Would it make sense to first fill the cache in `RegisterToWorkflow`? Thinking about some use case that has time sensitive workflow runs.

Reply:
Yeah this is what Ryan was referring to -- I'm going to hold off on doing this because stateful capabilities have more that can go wrong with this, but we can implement it if the metrics indicate we're spending a lot of time initializing instances.

Note that some capabilities take upwards of 60s, and this initialization should add at most 1 second, which is why I'm holding off.